infrastructure-formula/infrastructure/salt/git/formulas.sls

```diff
@@ -0,0 +1,20 @@
+{%- from 'infrastructure/salt/map.jinja' import git, formulas -%}
+
+{%- if 'repository' in formulas %}
+{%- set branch = formulas.get('branch', git.branch) %}
+
+salt_formula_clone:
+  git.latest:
+    - name: {{ formulas['repository'] }}
+    - target: {{ formulas.directory }}
+    - branch: {{ branch }}
+    - rev: {{ branch }}
+    {%- if not ( opts['test'] and not salt['user.info'](git.user) ) %}
```

Comment:
I remember someone saying `double-negatives are not hard to misunderstand.`

Originally I wanted to ask for an easier to understand rewrite, but I'm not sure if 

    {% if not opts['test'] or salt['user.info'](git.user) %}

(which should give the same result) would be easier to understand.

Maybe the best solution is a comment explaining the condition ;-) ("test mode fails if the user doesn't exist yet")

Reply:
Interesting, I did not consider using `or`, it took me a bit to get the logic work as expected hence I'd rather keep it.
I can add a comment though it does theoretically say in the commit message? :)

Reply:
Using `git blame` is a bit hard if you look at rpm-packaged formulas ;-) (and will also become harder if the line gets another minor change in the future). In comparison, a comment is easily available.

For the `or` - well, I did a small logic table to understand your logic, and then looked for alternative ways with the same result.

Reply:
Fair enough, I added it.

Would be interested in your logic table for future reference. :)

Reply:
One issue I see with the `or` (which might also be in my condition though) is that it will not render the `user` line in all non-test applies - even if the user exists.

Reply:
I'm not sure if I understand your problem.

Your code `if not ( opts['test'] and not  $whatever)` will in non-test mode be `if not ( False and not $whatever)` which is the same as `if not ( False )` (because "False and something" is always False). And finally, `if not ( False )` is effectively `if True`.

And that also means that in non-test mode, the `- user:` line will always be rendered.

Now let's check the same for my `or` variant `if not opts['test'] or $whatever`: in non-test mode, this will become `if not False or $whatever`, which means `if True or $whatever` - and `True or $whatever` is always `True` independent of the value of `$whatever`.

(Unfortunately I had the logic table on a scratch note which I didn't save, but it's not really hard to re-create, and might even help you to understand my comment, and why my variant also works.)

Reply:
Thanks for elaborating.
I think I understand your variant, but you seem to confirm my issue.
If test mode is `True`, then the user evaluation will _always_ be skipped. We do not want this - if the user exists, the test should also get the `user` line (otherwise the check whether there is an update to the repository will run as the user the Salt minion is running under).

Reply:
To be fair, it is only a minor issue.